### PR TITLE
Fix for invalid handle when calling dlopen twice

### LINKED ISSUE #8511: Calling dlopen twice for a non-existent module returns a handle the second time
Using the dynamic linking feature, executing the following code:

```c
void* handle1 = dlopen("invalid.wasm", 0);
void* handle2 = dlopen("invalid.wasm", 0);
```

has `handle2` non null. Internally, the structures contain a "loading" state that has not been cleaned up in `_dlopen` when the module has not been found on the first try.

### PR DESCRIPTION
**Pull request by jeromelaban**
Fixes #8511

Calling dlopen twice for an invalid module returns an invalid handle the second time.

### CODE REVIEW

File: tests/test_browser.py [hunk 1]
```diff
@@ -2408,6 +2408,16 @@ def test_preload_module(self):
...
+        void *invalid_handle = dlopen("/invalid.so", 0);
+        if (invalid_handle) {
+          REPORT_RESULT(4);
+          return 4;
+        }
+        void *invalid_handle2 = dlopen("/invalid.so", 0);
+        if (invalid_handle2) {
+          REPORT_RESULT(5);
+          return 5;
+        }
```
Comment (sbc100) [inline tests/test_browser.py]: Can you make this a separate test?    It doesn't seem related to preloading, right?

File: src/support.js [hunk 2]
```diff
@@ -302,14 +302,31 @@ function loadDynamicLibrary(lib, flags) {
...
+      })
+      .then(function (libModule) {
+        moduleLoaded(libModule);
+        return handle;
+      })
   }
 
-  moduleLoaded(getLibModule());
+  try {
+    moduleLoaded(getLibModule());
```
Comment (sbc100) [inline src/support.js]: It looks like just the `getLibModule()` call can be in the try block and the `moduleLoaded()` call can be moved below?   (At least that looks like it would be in line with what the async code is doing above.

File: src/support.js [hunk 3]
```diff
@@ -302,14 +302,31 @@ function loadDynamicLibrary(lib, flags) {
...
+
   if (flags.loadAsync) {
-    return getLibModule().then(function(libModule) {
-      moduleLoaded(libModule);
-      return handle;
-    })
+    return getLibModule()
+      .catch(function (e) {
+        cleanupOnError();
+        throw e;
```
Comment (sbc100) [inline src/support.js]: It doesn't look like this codepath is tested.  I'm also not knowledgeable about promises and how they interact with exceptions.    @kripken ?

Comment (kripken) [reply src/support.js]: The exceptions stuff here looks good to me.

**Comment by sbc100** [comment]
Can you add yourself to the AUTHORS file if you have not already.

**Comment by sbc100** [comment]
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.

**Comment by jeromelaban** [comment]
Thanks @sbc100. I still have to work on addressing those issues, though I have to familiarize myself with how to run tests locally first.
